README.md
## EJBCA PKI Server
This is an image for running your own [EJBCA PKI](https://www.ejbca.org) server.

[![build](https://circleci.com/gh/dataknightllc/docker-ejbca-image/tree/master.svg?style=shield&circle-token=f6ca71486c51971be9e2dbb5e07730c792b0aa89)](https://circleci.com/gh/dataknightllc/docker-ejbca-image/tree/master)

### Prerequisites
- Install [Docker Compose](https://github.com/docker/compose) on the system.
- Review the documentation for [EJBCA PKI](https://www.ejbca.org/docs/index.html) for details on how to configure and
  use the product.
- Choose a location on the system in which to store EJBCA PKI configuration settings and data.

### Container Setup
Follow the steps below to set up the container.  These instructions assume you have chosen **/opt/docker/ejbca** as
a base folder for storing settings and data.  If you choose a different path, please update the path in the commands
below accordingly.

1. On the Docker host, create the **docker-compose.yml** file inside the **/opt/docker/ejbca** folder.  The 
   format is shown below in the **Docker Compose Configuration** section.
2. Create the necessary configuration and data folders.
   - `host# mkdir -p /opt/docker/ejbca/{mysql,data}`
3. Pull the latest image from the registry:
   - `host# docker-compose pull ejbca`
4. Review the **Container Initialization** section below on how to pre-configure the container with your own custom
   settings, if desired.
5. Use **docker-compose** to bring up the container:
   - `host# docker-compose up ejbca`

In the future, you can update the container by simply re-running **docker-compose pull ejbca** followed by
**docker-compose up --force-recreate ejbca**.  These commands will automatically pull the latest version of
the image(s) from the registry and replace it without affecting your data or configurations.

### Docker Compose Configuration
Below is a sample `docker-compose.yml` file for your reference:
```
#
# Sample configuration for 'ejbca' container
#

version: "3"

services:
  mariadb:
    container_name: mariadb
    image: dataknightllc/mariadb:stable
    volumes:
      - /opt/docker/mariadb/rsyslog:/etc/rsyslog/conf.d
      - /opt/docker/mariadb/data:/var/lib/mysql
    ports:
      - 3306:3306
    environment:
      TZ: ${TZ:-UTC}
      SECRETS_FOLDER: /run/secrets
  
  ejbca:
    container_name: ejbca
    image: dataknightllc/ejbca:stable
    volumes:
      - /opt/docker/ejbca/syslog-ng:/etc/syslog-ng/conf.d
      - /opt/docker/ejbca/mysql:/var/lib/mysql
      - /opt/docker/ejbca/data:/var/lib/ejbca
    ports:
      - 8080:8080
      - 8442:8442
      - 8443:8443
    environment:
      TZ: ${TZ:-UTC}
```

### Backing Up the Container
Once you have configured the container per the instructions above, you simply need to backup the **/opt/docker/ejbca**
folder and subfolders to save your configuration and data.

### Container Initialization
This container image is configured so that the first time you boot it, it will automatically initialize a new
Certificate Authority and database for you with little to no intervention.  Please be aware that the first boot
of the container may take some time (~5 minutes or so) in order to initialize the MySQL database and Wildfly
server for running EJBCA.

You may notice several folders such as **/var/lib/mysql/.init** and **/var/lib/ejbca/.init** that are created
automatically during first boot.  You should not touch these folders or files.  They are used for keeping
initialization state for your container between reboots and upgrades or container rebuilds.  Likewise, once the
initial boot has been performed, you should not alter the values or settings in the **/var/lib/mysql/.vault** and
**/var/lib/ejbca/.vault** files unless otherwise noted.

If you wish to use the default settings, you do not have to do anything other than set up the container as described
in the steps above.  A default certificate authority called **Root Certificate Authority** will be created and 
initialized for you automatically during first boot.

If you wish to customize settings for the initialization, you may do so by following the instructions below.  Any
files created below should be owned by the **root** user and set to mode 0600.  Settings for _.properties_ files
described below are assigned values each on a single line using _setting=value_ syntax.  If the value of any setting
contains spaces, be sure to wrap it in double quote (") marks.  If you choose to create a file, it **must** contain
all of the settings specified.

**NOTE:** If you choose to supply your own password for any of the settings listed below, it must be a **minimum** of
6 characters long.

__MySQL User Customization__
If you wish to set your own password for the **root** and/or **ejbca** user, simply create a new file corresponding
to the username in the **/var/lib/mysql/.vault** folder.  The file should _only_ contain the password and nothing else.

__Certificate Authority Customization__
If you wish to customize the CA that is created during first boot, create a new file called **ca.properties** inside
the **/var/lib/ejbca/.vault** folder.
  - **ca.name**: The common name of the certificate authority (eg: "My CA")
  - **ca.dn**: The fully distinguished name of the certificate authority (eg: "CN=My CA,O=My Company,C=US") 
  - **ca.keytype**: Should always be **RSA**
  - **ca.keyspec**: The number of bits of encryption to use for the private key (eg: 1024 or 2048 or 4096)
  - **ca.signaturealgorithm**: Use **SHA256WithRSA** unless you know what you're doing
  - **ca.validity**: The number of days until the CA expires (eg: 7300 which is 20 years)
  - **ca.policy**: Use **null** unless you know what you're doing
  - **ca.tokenpassword**: The password used to encrypt the CA private key

__Java Customization__
If you wish to customize Java settings used during first boot, create a new file called **java.properties** inside
the **/var/lib/ejbca/.vault** folder.
  - **java.trustpassword**: The password used to encrypt the Java keystore for storing the SSL certificate for
                            the Wildfly web server.

__SuperAdmin Customization__
If you wish to customize settings used to create the super administrator account during first boot, create a new file
called **superadmin.properties** inside the **/var/lib/ejbca/.vault** folder.
  - **superadmin.cn**: The common name of the super administrator account (eg: admin)
  - **superadmin.dn**: The fully distinguished name of the super administrator account (eg: CN=admin)
  - **superadmin.password**: The password used to encrypt the PKCS12 keystore containing the user's SSL client
                             certificate
  - **superadmin.batch**: Must be set to **true**

__HTTPS Server Customization__
If you wish to customize settings used for the SSL certificate for the Wildfly web server generated during first boot,
create a new file called **httpsserver.properties** inside the **/var/lib/ejbca/.vault** folder.
  - **httpsserver.hostname**: The common name of the web server (eg: pki.mydomain.com)
  - **httpsserver.dn**: The fully distinguished name to use in the SSL certificate (eg: CN=pki.mydomain.com)
  - **httpsserver.password**: The password used to encrypt the private SSL key

__SMTP Server Customization__
If you wish to customize settings used for sending email from EJBCA, create a new file called **smtpserver.properties**
inside the **/var/lib/ejbca/.vault** folder.
  - **smtpserver.enabled**: **true** to enable and use the SMTP settings or **false** to ignore them
  - **smtpserver.port**: SMTP server port (eg: 25 or 587)
  - **smtpserver.host**: Hostname (or FQDN depending on your configuration) of the SMTP server
  - **smtpserver.from**: E-mail address used in the **From:** address
  - **smtpserver.user**: Username for SMTP server authentication or leave blank if not required
  - **smtpserver.password**: Password for SMTP server authentication or leave blank if not required
  - **smtpserver.use_tls**: **true** to use TLS when connecting to the SMTP server or **false** to use cleartext

### Logging into EJBCA as a Super Administrator
Once the server is up and running, you'll want to connect to the Web UI and login as the administrator.

1. Download the **/var/lib/ejbca/p12/superadmin.p12** file from the container to your local system.
2. Grab the value of **superadmin.password** from the **/var/lib/ejbca/.vault/superadmin.properties** file in the
   container.
3. Import the PKCS12 file onto your system (for IE, Edge and Chrome) or into your browser (for Firefox).
4. Restart your browser.
5. Navigate to https://ejbcaserver:8443/ejbca/adminweb where _ejbcaserver_ is the hostname or IP of your 
   EJBCA server.

### Activating the Certificate Authority
After the initial boot of the container, you will need to activate the certificate authority for use.  You'll need to
log into the Web UI and perform the steps that follow.

1. Open your browser and navigate to https://ejbcaserver:8443/ejbca/adminweb where _ejbcaserver_ is the hostname or IP
   of your EJBCA server.
2. Click the **CA Activation** link under **CA Functions**.
3. Under the **Crypto Token Action** check the **Activate** box.
4. Grab the value of **ca.tokenpassword** from the **/var/lib/ejbca/.vault/ca.properties** file in the container and
   enter it next to **Crypto Token activation code**.
5. Click the **Apply** button.

Note that afterwards you may want to turn on auto-activation by clicking on the link for the CA and going into edit
mode and making the change.

### Changing the Certificate Authority Password
...TO BE DOCUMENTED...

### Updating the Wildfly Web Server Certificate
If you wish to update the SSL certificate used by the Wildfly front-end web server because it has expired or if you
wish to change its properties or use a different CA to sign the certificate, you'll need to perform the steps that
follow.

__Using the EJBCA Certificate Authority__
Follow the steps below to generate a new certificate using the EJBCA Certificate Authority you created.

1. Open your browser and navigate to https://ejbcaserver:8443/ejbca/adminweb where _ejbcaserver_ is the hostname or IP
   of your EJBCA server.
2. Click the **Search End Entities** link under **RA Functions**.
3. Next to **Search end entity with username** enter **tomcat** and click **Search**.
4. Click the **Edit End Entity** link in the search result.
5. Change the **Status** from **Generated** to **New**.
6. Grab the value of **httpsserver.password** from the **/var/lib/ejbca/.vault/httpsserver.properties** file in the 
   container and enter it in the **Password (or Enrollment Code)** and **Confirm Password** fields.
7. Finish editing the entity as desired and click **Save** when complete.
8. Run the following command from a shell within the container:
   - `/opt/ejbca/bin/ejbca.sh batch`
   - `keytool -changealias -keystore /var/lib/ejbca/p12/tomcat.jks -alias _CertCN_ -destalias ejbca`
     (Be sure to change _CertCN_ to the actual CN you specified for the certificate!)
   - `cp /var/lib/ejbca/wildfly/keystore/keystore.jks /var/lib/ejbca/wildfly/keystore/keystore.jks.bak`
   - `cp /var/lib/ejbca/p12/tomcat.jks /var/lib/ejbca/wildfly/keystore/keystore.jks`
   - `chown wildfly:wildfly /var/lib/ejbca/wildfly/keystore/keystore.jks*`
   - `/opt/wildfly/bin/jboss-cli.sh -c --command=":shutdown(restart=true)"`

__Using a 3rd Party Certificate Authority__
Follow the steps below to generate a new certificate using a 3rd party CA such as Verisign, Comodo, Thawte, etc.

1. Open a shell on the container and run the following command:  
   `keytool -genkey -alias ejbca -keyalg RSA -keysize 2048 -keystore /tmp/keystore.jks`
2. Grab the value of **httpsserver.password** from the **/var/lib/ejbca/.vault/httpsserver.properties** file in the 
   container and use it when prompted for a password for the keystore.
3. Enter the details for the certificate as you are prompted for them.
4. Press **ENTER** when asked for the password for the **ejbca** key.
5. Now run the following command:
   `keytool -certreq -alias ejbca -keyalg RSA -file /tmp/request.csr -keystore /tmp/keystore.jks`
6. Submit the **/tmp/request.csr** file to the CA for signing.  Be sure to request a certificate for Java/Tomcat/JBOSS
   so that the signed certificate is in PKCS7 (.p7b) format.
7. Upload the signed PKCS7 bundle to the container as **/tmp/cert.p7b**.
8. Now run the following commands:
   - `keytool -import -trustcacerts -alias ejbca -file /tmp/cert.p7b -keystore /tmp/keystore.jks`
   - `cp /var/lib/ejbca/wildfly/keystore/keystore.jks /var/lib/ejbca/wildfly/keystore/keystore.jks.bak`
   - `cp /tmp/keystore.jks /var/lib/ejbca/wildfly/keystore/keystore.jks`
   - `chown wildfly:wildfly /var/lib/ejbca/wildfly/keystore/keystore.jks*`
   - `/opt/wildfly/bin/jboss-cli.sh -c --command=":shutdown(restart=true)"`

__Importing an Existing SSL Key Pair__
...TO BE DOCUMENTED...

If you encounter any issues with Wildfly not restarting, simply copy the **keystore.jks.bak** file back to
**keystore.jks** and issue the restart command to **jboss-cli.sh**.

### Troubleshooting Errors
If you have issues, be sure to check the log files in the **/var/log** folder for details.  The majority of information
will be stored in the **wildfly.log** file.

### Additional Help or Questions
If you have questions, find bugs or need additional help, please send an email to 
[support@dataknight.co](mailto:support@dataknight.co).


CHANGELOG.md
Release History:
================

**6.5.0-r1** _(Released 06/17/2017)_
- Separated out MySQL server
- Changed base image to **dataknightllc/wildfly:stable**
- Changed initialization scripts to use secrets folder
<br /><br />

**6.5.0-r0** _(Released 06/02/2017)_
- Initial release of the container image
- EJBCA Community version 6.5.0.5


files/usr/local/lib/mysql.sh
#
# Shared functions and variables for MySQL server
#

# Setup shared variables
export MYSQL_DATADIR="/var/lib/mysql"
export MYSQL_INITDIR="${MYSQL_DATADIR}/.init"
export MYSQL_VAULTDIR="${MYSQL_DATADIR}/.vault"
export MYSQL_LOG="/var/log/mysqld_safe.log"
export MYSQL_INIT_LOG="/var/log/init_mysql_server.log"

#
# mysql_init()
#   Initializes MySQL database if it has not been set up already
#
mysql_init() {
  # Already initialized
  if [ -f "${MYSQL_INITDIR}/ready" ]; then
    return 0
  fi
  echo "Initializing MySQL server..."

  # Update folder permissions
  if [ ! -d "/run/mysqld" ]; then
    mkdir -p /run/mysqld 
    chown mysql:mysql /run/mysqld
  fi
  chown -hR mysql:mysql "${MYSQL_DATADIR}"

  # Create init dir
  if [ ! -d "${MYSQL_INITDIR}" ]; then
    mkdir -p "${MYSQL_INITDIR}"
  fi
  chown -hR root:root "${MYSQL_INITDIR}"

  # Create vault dir
  if [ ! -d "${MYSQL_VAULTDIR}" ]; then
    mkdir -p "${MYSQL_VAULTDIR}"
  fi
  chown -hR root:root "${MYSQL_VAULTDIR}"
  chmod 750 "${MYSQL_VAULTDIR}"

  # Set user passwords
  for user in root ejbca; do
    if [ ! -f "${MYSQL_VAULTDIR}/${user}" ]; then
      echo $(openssl rand 18 -base64) > "${MYSQL_VAULTDIR}/${user}"
    fi
    chown 0:0 "${MYSQL_VAULTDIR}/${user}"
    chmod 600 "${MYSQL_VAULTDIR}/${user}"
  done

  # Create temp startup script
  script=$(mktemp)
  cat << EOF > "${script}"
DROP DATABASE test;

USE mysql;
DELETE FROM user WHERE user != 'root' OR host != 'localhost';
UPDATE user SET password=PASSWORD("$(cat "${MYSQL_VAULTDIR}/root")") WHERE user = 'root' AND host = 'localhost'; 

CREATE DATABASE IF NOT EXISTS ejbca CHARACTER SET utf8 COLLATE utf8_general_ci;
# We cannot do a GRANT in a bootstrap, so this gets executed when starting the server
# GRANT ALL PRIVILEGES ON ejbca.* TO 'ejbca'@'localhost' IDENTIFIED BY "${MYSQL_EJBCA_PASSWORD}";
FLUSH PRIVILEGES;
EOF

  # Initialize the server
  /usr/bin/mysql_install_db --user=mysql > "${MYSQL_INIT_LOG}" 2>&1
  /usr/bin/mysqld --user=mysql --bootstrap --verbose=0 < "${script}" >> "${MYSQL_INIT_LOG}" 2>&1
  rm -f "${script}"

  # Complete
  echo "  --> MySQL server initialized. Check the '${MYSQL_INIT_LOG}' file for details."
  touch "${MYSQL_INITDIR}/ready"
  return 0
}

#
# mysql_init_users()
#   Initializes required MySQL user accounts
#
mysql_init_users() {
  # Already initialized
  if [ -f "${MYSQL_INITDIR}/users" ]; then
    return 0
  fi
  echo "Initializing MySQL users..."

  # Configure 'ejbca' user
  root_pass="$(cat "${MYSQL_VAULTDIR}/root")"
  ejbca_pass="$(cat "${MYSQL_VAULTDIR}/ejbca")"
  mysql --user=root --password="${root_pass}" -e \
      "GRANT ALL PRIVILEGES ON ejbca.* TO 'ejbca'@'localhost' IDENTIFIED BY '${ejbca_pass}';" >>"${MYSQL_INIT_LOG}" 2>&1

  # Complete
  echo "  --> MySQL users initialized. Check the '${MYSQL_INIT_LOG}' file for details."
  touch "${MYSQL_INITDIR}/users"
  return 0
}

#
# mysql_start()
#   Starts the MySQL server
#
mysql_start() {
  # Initialize DB
  mysql_init

  # Start server
  echo "Starting MySQL server..."
  /usr/bin/mysqld_safe --user=mysql >"${MYSQL_LOG}" 2>&1 &
  mysql_wait 60
  echo "  --> MySQL server started. Check the '${MYSQL_LOG}' file for details."

  # Initialize users
  mysql_init_users
  return 0
}

#
# mysql_wait(num_sec)
#   Waits up to 'num_sec' seconds for MySQL server socket become available
#
mysql_wait() {
  max_wait=$1

  delay=1
  total_wait=0
  until [ $total_wait -gt $max_wait ]; do
    if [ -S /run/mysqld/mysqld.sock ]; then
      break
    fi
    echo "[${total_wait}/${max_wait}s elapsed] Waiting for MySQL server..."
    sleep $delay
    let total_wait=$total_wait+$delay
  done
  if [ $total_wait -gt $max_wait ]; then
    return 1
  else
    return 0
  fi
}


files/usr/local/lib/ejbca.sh
#
# Shared functions and variables for EJBCA
#

# Setup shared variables
export EJBCA_HOME="/opt/ejbca"
export EJBCA_DATADIR="/var/lib/ejbca"
export PATH=$PATH:${EJBCA_HOME}/bin

#
# ejbca_create_ca()
#   Performs creation of the root certificate authority
#
ejbca_create_ca() {
  # Already initialized
  if [ -f "${EJBCA_DATA_INITDIR}/ca" ]; then
    return 0
  fi
  echo "  Creating certificate authority..."

  # Configure properties file

    echo 'ca.name="Root Certificate Authority"' >"${ca_properties_file}"
    echo 'ca.dn="CN=Root Certificate Authority"' >>"${ca_properties_file}"
    echo 'ca.keytype=RSA' >>"${ca_properties_file}"
    echo 'ca.keyspec=2048' >>"${ca_properties_file}"
    echo 'ca.signaturealgorithm=SHA256WithRSA' >>"${ca_properties_file}"
    echo 'ca.validity=7300' >>"${ca_properties_file}"
    echo 'ca.policy=null' >>"${ca_properties_file}"
    echo "ca.tokenpassword=\"$(openssl rand 16 -base64)\"" >>"${ca_properties_file}"

    echo "java.trustpassword=\"$(openssl rand 16 -base64)\"" >"${java_properties_file}"

    echo 'superadmin.cn=admin' >"${superadmin_properties_file}"
    echo 'superadmin.dn="CN=admin"' >>"${superadmin_properties_file}"
    echo "superadmin.password=\"$(openssl rand 16 -base64)\"" >>"${superadmin_properties_file}"
    echo 'superadmin.batch=true' >>"${superadmin_properties_file}"

    echo 'httpsserver.hostname=ejbca' >"${httpsserver_properties_file}"
    echo 'httpsserver.dn="CN=ejbca"' >>"${httpsserver_properties_file}"
    echo "httpsserver.password=\"$(openssl rand 16 -base64)\"" >>"${httpsserver_properties_file}"

    echo 'smtpserver.enabled=false' >"${smtpserver_properties_file}"
    echo 'smtpserver.port=25' >>"${smtpserver_properties_file}"
    echo 'smtpserver.host=localhost' >>"${smtpserver_properties_file}"
    echo 'smtpserver.from=ejbca-noreply@localhost' >>"${smtpserver_properties_file}"
    echo 'smtpserver.user=' >>"${smtpserver_properties_file}"
    echo 'smtpserver.password=' >>"${smtpserver_properties_file}"
    echo 'smtpserver.use_tls=false' >>"${smtpserver_properties_file}"




#(cli) ca init &quot;${ca.name}&quot; &quot;${ca.dn}&quot; ${ca.tokentype} ${ca.tokenpassword} ${ca.keyspec} ${ca.keytype} ${ca.validity} ${ca.policy} ${ca.signaturealgorithm} ${install.catoken.command} ${install.certprofile.command} -superadmincn &quot;${superadmin.cn}&quot;"
#(cli) ra addendentity tomcat --password ${httpsserver.password} &quot;${httpsserver.dn}&quot; --altname &quot;${httpsserver.an}&quot; &quot;${ca.name}&quot; 1 JKS --certprofile SERVER
#(cli) ra setclearpwd tomcat ${httpsserver.password}
#(cli) batch tomcat

#(cli) ra addendentity superadmin --password ${superadmin.password} &quot;${superadmin.dn}&quot; &quot;${ca.name}&quot; 1 ${superadmin.keystoretype}"
#(cli) ra setclearpwd superadmin ${superadmin.password}
#(cli) batch superadmin

#(cli) ca getcacert &quot;${ca.name}&quot; ${java.io.tmpdir}/rootca.der -der
#(keytool) -v -alias &quot;${ca.name}&quot; -import -trustcacerts -file '${java.io.tmpdir}/rootca.der' -keystore '${trust.keystore}' -storepass ${trust.password} -noprompt
#rm rootca.der


  # Run the script to create the CA
  script=$(mktemp)
  chmod a+rx "${script}"
  echo '#!/bin/bash' >"${script}"
  flags=""
  for f in $(find "${EJBCA_SECRETDIR}" -type f -name \*.properties); do
    flags="${flags} $(cat "$f" | awk 'NF {print "-D" $0}')"
  done
  echo -n "ant runinstall " >>"${script}"
  echo $flags >>"${script}"
  su-exec wildfly "${script}" >>"${EJBCA_LOG}" 2>&1
  rm -f "${script}"

  # Deploy the keystore to Wildfly
  echo "  Deploying keystore..."
  su-exec wildfly ant deploy-keystore >>"${EJBCA_LOG}" 2>&1
  echo "    --> Keystore deployed to Wildfly."
  
  # Complete
  echo "    --> Certificate Authority has been configured."
  touch "${EJBCA_DATA_INITDIR}/ca"
  return 0
}

#
# ejbca_init()
#   Initializes and deploys EJBCA if it has not been set up already
#
ejbca_init() {
  echo "Initializing EJBCA server..."

  ejbca_startup_check
  ejbca_init_vars
  wildfly_init
  wildfly_start_bg
  #ejbca_create_ca
  wildfly_stop

  echo "EJBCA server initialized."
  return 0
}

#
# ejbca_init_vars()
#   Initializes all EJBCA environment variables
#
ejbca_init_vars() {
  # Set up MySQL variables
  if [[ -z "${MYSQL_CONNECTION_URL}" ]]; then
    export MYSQL_CONNECTION_URL="mysql://mariadb:3306/ejbca"
  fi
  if [[ -z "${MYSQL_USERNAME}" ]]; then
    if [[ -z "${MYSQL_USERNAME_SECRET}" ]]; then
      export MYSQL_USERNAME="ejbca"
    else
      ejbca_read_secret MYSQL_USERNAME_SECRET MYSQL_USERNAME
    fi
  fi
  if [[ -z "${MYSQL_PASSWORD}" ]]; then
    if [[ -z "${MYSQL_PASSWORD_SECRET}" ]]; then
      echo "MYSQL_PASSWORD or MYSQL_PASSWORD_SECRET environment variable must be set."
      exit 1
    else
      ejbca_read_secret MYSQL_PASSWORD_SECRET MYSQL_PASSWORD
    fi
  fi

  # Setup WildFly variables
  if [[ -z "${WILDFLY_SERVER_CN}" ]]; then
    export WILDFLY_SERVER_CN="ejbca"
  fi
  if [[ -z "${WILDFLY_SERVER_ALT_NAMES}" ]]; then
    export WILDFLY_SERVER_ALT_NAMES=""
  fi
  if [[ -z "${WILDFLY_STOREPASS}" ]]; then
    if [[ -z "${WILDFLY_STOREPASS_SECRET}" ]]; then
      export WILDFLY_STOREPASS="changeit"
    else
      ejbca_read_secret WILDFLY_STOREPASS_SECRET WILDFLY_STOREPASS
    fi
  fi
  if [[ -z "${WILDFLY_TRUSTSTOREPASS}" ]]; then
    if [[ -z "${WILDFLY_TRUSTSTOREPASS_SECRET}" ]]; then
      export WILDFLY_TRUSTSTOREPASS="changeit"
    else
      ejbca_read_secret WILDFLY_TRUSTSTOREPASS_SECRET WILDFLY_TRUSTSTOREPASS
    fi
  fi

  # Setup SMTP server variables
  if [[ -z "${SMTPSERVER_ENABLED}" ]]; then
    export SMTPSERVER_ENABLED="false"
  fi
  if [[ "${SMTPSERVER_ENABLED}" == "true" ]]; then
    if [[ -z "${SMTPSERVER_FROM}" ]]; then
      export SMTPSERVER_FROM="ejbca-noreply@ejbca"
    fi
    if [[ -z "${SMTPSERVER_USE_TLS}" ]]; then
      export SMTPSERVER_USE_TLS="false"
    fi
    if [[ -z "${SMTPSERVER_HOST}" ]]; then
      export SMTPSERVER_HOST="smtp"
    fi
    if [[ -z "${SMTPSERVER_PORT}" ]]; then
      export SMTPSERVER_PORT=25
    fi
    if [[ -z "${SMTPSERVER_AUTH_REQUIRED}" ]]; then
      export SMTPSERVER_AUTH_REQUIRED="false"
    fi
    if [[ "${SMTPSERVER_AUTH_REQUIRED}" == "true" ]]; then
      if [[ -z "${SMTPSERVER_USERNAME}" ]]; then
        if [[ -z "${SMTPSERVER_USERNAME_SECRET}" ]]; then
          export SMTPSERVER_USERNAME="ejbca"
        else
          ejbca_read_secret SMTPSERVER_USERNAME_SECRET SMTPSERVER_USERNAME
        fi
      fi
      if [[ -z "${SMTPSERVER_PASSWORD}" ]]; then
        if [[ -z "${SMTPSERVER_PASSWORD_SECRET}" ]]; then
          echo "SMTPSERVER_PASSWORD or SMTPSERVER_PASSWORD_SECRET environment variable must be set."
          exit 1
        else
          ejbca_read_secret SMTPSERVER_PASSWORD_SECRET SMTPSERVER_PASSWORD
        fi
      fi
    fi
  fi

  # Setup CA server variables
  if [[ -z "${CA_NAME}" ]]; then
    export CA_NAME="Root Certificate Authority"
  fi
  if [[ -z "${CA_DN}" ]]; then
    export CA_DN="CN=Root Certificate Authority"
  fi
  if [[ -z "${CA_KEYTYPE}" ]]; then
    export CA_KEYTYPE="RSA"
  fi
  if [[ -z "${CA_KEYSPEC}" ]]; then
    export CA_KEYSPEC=2048
  fi
  if [[ -z "${CA_SIGNATUREALGORITHM}" ]]; then
    export CA_SIGNATUREALGORITHM="SHA256WithRSA"
  fi
  if [[ -z "${CA_VALIDITY}" ]]; then
    export CA_VALIDITY=7300
  fi
  if [[ -z "${CA_POLICY}" ]]; then
    export CA_POLICY="null"
  fi
  if [[ -z "${CA_PASSWORD}" ]]; then
    if [[ -z ${CA_PASSWORD_SECRET} ]]; then
      export CA_PASSWORD="changeit"
    else
      ejbca_read_secret CA_PASSWORD_SECRET CA_PASSWORD
    fi
  fi

  # Setup SuperAdmin user variables
  if [[ -z "${SUPERADMIN_CN}" ]]; then
    if [[ -z ${SUPERADMIN_CN_SECRET} ]]; then
      export SUPERADMIN_CN="admin"
    else
      ejbca_read_secret SUPERADMIN_CN_SECRET SUPERADMIN_CN
    fi
  fi
  if [[ -z "${SUPERADMIN_DN}" ]]; then
    if [[ -z ${SUPERADMIN_DN_SECRET} ]]; then
      export SUPERADMIN_DN="CN=admin"
    else
      ejbca_read_secret SUPERADMIN_DN_SECRET SUPERADMIN_DN
    fi
  fi
  if [[ -z "${SUPERADMIN_PASSWORD}" ]]; then
    if [[ -z ${SUPERADMIN_PASSWORD_SECRET} ]]; then
      export SUPERADMIN_PASSWORD="changeit"
    else
      ejbca_read_secret SUPERADMIN_PASSWORD_SECRET SUPERADMIN_PASSWORD
    fi
  fi
  return 0  
}

#
# ejbca_read_secret($secret, $store_as)
#   Reads the file that the given $secret environment variable points to and return the value as the $store_as variable
#
ejbca_read_secret() {
  secret=$1
  store_as=$2

  # Make sure the file exists
  secret_file=$(eval "echo \$${secret}")
  if [ ! -f "${secret_file}" ]; then
    echo "Cannot find ${secret} file '${secret_file}'."
    exit 1
  fi

  # Store the value
  value=$(cat "${secret_file}")
  eval "${store_as}=\"${value}\""
  return 0
}

#
# ejbca_start()
#   Starts the EJBCA server
#
ejbca_start() {
  wildfly_start
}

#
# ejbca_startup_check()
#   Ensures folders and files required to start EJBCA exist and have the correct permissions
#
ejbca_startup_check() {
  # Make sure folders exist
  if [ ! -d "${EJBCA_DATADIR}/conf" ]; then
    mkdir -p "${EJBCA_DATADIR}/conf"
    find "${EJBCA_HOME}/conf.dist" -name \*.properties | xargs cp -t "${EJBCA_DATADIR}/conf/"
  fi
  for dir in keystore conf/logdevices conf/plugins; do
    if [ ! -d "${EJBCA_DATADIR}/${dir}" ]; then
      mkdir -p "${EJBCA_DATADIR}/${dir}"
    fi
  done

  # Make sure permissions are correct
  chown -hR wildfly:wildfly "${EJBCA_DATADIR}"
  chmod 0750 "${EJBCA_DATADIR}"
  return 0
}

#
# ejbca_update_config()
#   Regenerates the WildFly standalone.xml configuration file based on the current environment variables
#
ejbca_update_config() {
  echo "Regenerating WildFly standalone.xml file..."
  ejbca_startup_check
  wildfly_startup_check
  ejbca_init_vars
  wildfly_create_config
  echo "A new standalone.xml file has been generated."
  return 0
}
